### PR TITLE
JP-3669: 'OLS' Conditional Read Noise Recalculation for CHARGELOSS

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3669](https://jira.stsci.edu/browse/JP-3669)

This PR updates the code flow for read noise calculation.  For ramps that have CHARGELOSS flagging, when calling the "OLS_C" algorithm, this recalculation is done in the C-extension in the ramp fitting code in STCAL, so the step code calculations defined in the JWST step code no longer needs to be run and will result in errors if it is run.

The step code that handles the recalculation of the read noise for CHARGELOSS flags is now only done when the "OLS" algorithm is used.

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by kmacdonald-stsci** [comment]
The regression testing using "-k rate" filter in the new github actions regression testing is here:

https://github.com/spacetelescope/RegressionTests/actions/runs/10309026450

This has expected failures and differences.  A full regression test is currently being run on Jenkins: 

https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1639/

**Review by melanieclarke — COMMENTED** [review]
This looks fine to me.  

The regression tests show a couple pixels different for NIRISS readnoise variances.  Was this running with OLS_C and pointing to your stcal branch?  Have you tracked down the differences?

File: CHANGES.rst
Comment (emolter) [inline CHANGES.rst:62]:
```suggestion
  fitting algorithm, this recalculation is being done in the C-extension, so
```

**Review by emolter — APPROVED** [review]
LGTM, assuming the chargeloss calculation is indeed handled in the C code - I didn't check myself.

I agree with Melanie RE the changelog entry.

**Comment by kmacdonald-stsci** [comment]
> This looks fine to me.
> 
> The regression tests show a couple pixels different for NIRISS readnoise variances. Was this running with OLS_C and pointing to your stcal branch? Have you tracked down the differences?

Yes, the regression test using `OLS_C`.  The differences are known and expected, due to the known segmentation bug in the `OLS` code.

File: CHANGES.rst
Comment (melanieclarke) [inline CHANGES.rst:68]: Thanks for the changes!  But I think this could still be more concise:

```suggestion
- Moved the read noise variance recalculation for CHARGELOSS flagging to the C
  implementation, when the algorithm is ``OLS_C``. [#8697, spacetelescope/stcal#278]

```

**Review by drlaw1558 — APPROVED** [review]
Linked to https://github.com/spacetelescope/stcal/pull/275

Looks good to me!  I've successfully tested that:

1) Runtime improves as expected in test case jw02079004003_03101_00001_nis.  In the old version it takes about 2.5 minutes to run the relevant piece of the RN recalculation, while with this PR it takes about 2.5 seconds.

2) RN variances stay as expected in test case jw01510001001_02108_00001_nis using both the OLS and OLS_C algorithms.

3) Checked that the SCI and variance arrays did not change for other instruments that don't use charge migration.

File: CHANGES.rst
Comment (melanieclarke) [inline CHANGES.rst:89]: Looking at #8613, I think this comment was supposed to go under the outlier detection heading.

Comment (kmacdonald-stsci) [reply CHANGES.rst:89]: This isn't my change.  That bullet is under `set_telescope_pointing` in the master repo.  It only appears as my edit because of git's problem with CHANGES.rst always claiming merge conflicts:

https://github.com/spacetelescope/jwst/blob/9667e6c8fcdf316aa558544afa54c7b8528a5540/CHANGES.rst?plain=1#L70

Comment (melanieclarke) [reply CHANGES.rst:89]: Thanks, it should be fixed on main now.

**Review by melanieclarke — APPROVED** [review]
I think this should be good to go, given David's tests and the clean regression tests.  I'll approve now, but it does need a conflict fixed in the change log and the stcal change needs to be merged first.  When that's done, go ahead and merge this one in too.

**Comment by tapastro** [comment]
Running regression tests before merging: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1701/

**Pull request merged**
